File: slidingWindow/longestPalindromicSubstring.js
// write a function that, given a string, returns its longest palindromic substring.
// Note that sing-character strings are palindromes.

// their solution:
function longestPalindromicSubstring(string) {
  const currentLongest = [0, 1];
  for (let i = 1; i < string.length; i++) {
    const odd = getLongestPalindromeFrom(string, i - 1, i + 1);
    const even = getLongestPalindromeFrom(string, i - 1, i);
    const longest = odd[1] - odd[0] > even[1] - even[0] ? odd : even;
    currentLongest = currentLongest[1] - currentLongest[0] > longest[1] - longest[0] ? currentLongest : longest;
  }
  return string.slice(currentLongest[0], currentLongest[1]);

  function getLongestPalindromeFrom(string, leftIdx, rightIdx) {
    while (leftIdx >= 0 && rightIdx < string.length) {
      if (string[leftIdx] !== string[rightIdx]) {
        break;
      }
      leftIdx--;
      rightIdx++;
    }
    return [leftIdx + 1, rightIdx];
  }
}

// my solution, which is correct, but more expensive complexity-wise
function longestPalindromicSubstring(string) {
  if (!string) {
    return null;
  }
  if (string.length === 1) {
    return string;
  }
  let palindrome = string[0];

  function palindromeCheck(str) {
    let start = 0;
    let end = str.length - 1;
    while (start <= end) {
      if (str[start] === str[end]) {
        start++;
        end--;
      } else {
        return false;
      }
    }
    return true;
  }

  for (let i = 0; i < string.length; i++) {
    for (let j = i + 1; j < string.length; j++) {
      let current = string.slice(i, j + 1);
      if (palindromeCheck(current) === true && current.length > palindrome.length) {
        palindrome = current;
      }
    }
  }
  return palindrome;
}

let s = "axyzzyxf"
console.log(longestPalindromicSubstring(s));

File: slidingWindow/longestPalindromicSubstring.py
# my easier to understand version
def longestPalindromicSubstring(string):
    if not string:
        return None

    length = len(string)
    palindrome = string[0]

    def palindromeCheck(str):
        start = 0
        end = len(str) - 1
        while start <= end:
            if str[start] == str[end]:
                start += 1
                end -= 1
            else:
                return False
        return True

    for i in range(0, length):
        for j in range(i + 1, length):
            current = string[i:j + 1]
            if palindromeCheck(current) == True and len(current) > len(palindrome):
                palindrome = current

    return palindrome


# version from algoexpert
def longestPalindromicSubstring(string):
    currentLongest = [0, 1]
    # eventually we will slice out the palindrome from the string, i.e. string[0:1]
    # we start at 1 because we already know that the first character counts as palindrome
    for i in range(1, len(string)):
        # i - 1 is the left index, i + 1 is the right index
        odd = getLongestPalindromeFrom(string, i - 1, i + 1)
        # we use i because that is the starting index
        even = getLongestPalindromeFrom(string, i - 1, i)
        # somehow this key thing checks to see which one is longer?
        longest = max(odd, even, key=lambda x: x[1] - x[0])
        currentLongest = max(longest, currentLongest,
                             key=lambda x: x[1] - x[0])
    return string[currentLongest[0]:currentLongest[1]]


def getLongestPalindromeFrom(string, leftIdx, rightIdx):
    while leftIdx >= 0 and rightIdx < len(string):
        if string[leftIdx] != string[rightIdx]:
            break
        leftIdx -= 1
        rightIdx += 1
    return [leftIdx + 1, rightIdx]
